WordNet Summary:

WordNet is a lexical database of nouns, verbs, adjectives and adverbs that provides short definitions called glosses, and use examples.  WordNet links words into semantic relations including synonyms, hyponyms, meronyms, etc.

In [ ]:
import nltk
nltk.download('all')

In [ ]:
from nltk.corpus import wordnet as wn
from nltk.wsd import lesk
from nltk.corpus import sentiwordnet as swn

In [ ]:
# output synsets of the database (noun)
wn.synsets('relation')

[Synset('relation.n.01'),
 Synset('sexual_intercourse.n.01'),
 Synset('relative.n.01'),
 Synset('relation.n.04'),
 Synset('relation_back.n.01'),
 Synset('relation.n.06')]

In [ ]:
# Select one sysnset from the list of synsets
# Extract its definition, usage examples and lemmas.

print("definition: " + wn.synset('sexual_intercourse.n.01').definition())
print(wn.synset('sexual_intercourse.n.01').examples())
print(wn.synset('sexual_intercourse.n.01').lemmas())


definition: the act of sexual procreation between a man and a woman; the man's penis is inserted into the woman's vagina and excited until orgasm and ejaculation occur
[]
[Lemma('sexual_intercourse.n.01.sexual_intercourse'), Lemma('sexual_intercourse.n.01.intercourse'), Lemma('sexual_intercourse.n.01.sex_act'), Lemma('sexual_intercourse.n.01.copulation'), Lemma('sexual_intercourse.n.01.coitus'), Lemma('sexual_intercourse.n.01.coition'), Lemma('sexual_intercourse.n.01.sexual_congress'), Lemma('sexual_intercourse.n.01.congress'), Lemma('sexual_intercourse.n.01.sexual_relation'), Lemma('sexual_intercourse.n.01.relation'), Lemma('sexual_intercourse.n.01.carnal_knowledge')]


In [ ]:
# Traversing the hierachy of selected synset
relations = wn.synset('sexual_intercourse.n.01')
hyp = relations.hypernyms()[0]
top = wn.synset('entity.n.01')
while relations:
  print(hyp)
  if hyp == top:
    break
  if hyp.hypernyms():
    hyp = hyp.hypernyms()[0]


Synset('sexual_activity.n.01')
Synset('bodily_process.n.01')
Synset('organic_process.n.01')
Synset('process.n.06')
Synset('physical_entity.n.01')
Synset('entity.n.01')


In a WordNet hierachy for nouns. the top-most noun level is called entity. Each sysnset has a hypernym and hyponym relationship. The top-level of the noun hierachy consists of abstract concepts while the lowe level synsets represent more specific categories.

In [ ]:
# print hypernyms, hyponyms, meronyms, holonyms, antonym
print(wn.synset('sexual_intercourse.n.01').hypernyms())
print(wn.synset('sexual_intercourse.n.01').hyponyms())

[Synset('sexual_activity.n.01')]
[Synset('defloration.n.02'), Synset('fuck.n.01'), Synset('hank_panky.n.01'), Synset('penetration.n.06'), Synset('unlawful_carnal_knowledge.n.01')]


In [ ]:
# output meronyms
wn.synset('sexual_intercourse.n.01').part_meronyms()

[Synset('insemination.n.02')]

In [ ]:
wn.synset('sexual_intercourse.n.01').part_holonyms()

[]

In [ ]:
from nltk.corpus.reader.wordnet import Synset
#output antonyms
relations = wn.synsets('relations', pos=wn.NOUN)[2]
Synset('sexual_intercourse.n.01')
relations.lemmas()
relations.lemmas()[0].antonyms()

[]

Select Verb

In [ ]:
# Select verb and output all synsets
wn.synsets('run')

[Synset('run.n.01'),
 Synset('test.n.05'),
 Synset('footrace.n.01'),
 Synset('streak.n.01'),
 Synset('run.n.05'),
 Synset('run.n.06'),
 Synset('run.n.07'),
 Synset('run.n.08'),
 Synset('run.n.09'),
 Synset('run.n.10'),
 Synset('rivulet.n.01'),
 Synset('political_campaign.n.01'),
 Synset('run.n.13'),
 Synset('discharge.n.06'),
 Synset('run.n.15'),
 Synset('run.n.16'),
 Synset('run.v.01'),
 Synset('scat.v.01'),
 Synset('run.v.03'),
 Synset('operate.v.01'),
 Synset('run.v.05'),
 Synset('run.v.06'),
 Synset('function.v.01'),
 Synset('range.v.01'),
 Synset('campaign.v.01'),
 Synset('play.v.18'),
 Synset('run.v.11'),
 Synset('tend.v.01'),
 Synset('run.v.13'),
 Synset('run.v.14'),
 Synset('run.v.15'),
 Synset('run.v.16'),
 Synset('prevail.v.03'),
 Synset('run.v.18'),
 Synset('run.v.19'),
 Synset('carry.v.15'),
 Synset('run.v.21'),
 Synset('guide.v.05'),
 Synset('run.v.23'),
 Synset('run.v.24'),
 Synset('run.v.25'),
 Synset('run.v.26'),
 Synset('run.v.27'),
 Synset('run.v.28'),
 Synset('run.v.

In [ ]:
# Select one sysnset from the list of synsets
# Extract its definition, usage examples and lemmas.

print(wn.synset('run.v.04').definition())
print(wn.synset('run.v.04').examples())
print(wn.synset('run.v.04').lemmas())

direct or control; projects, businesses, etc.
['She is running a relief operation in the Sudan']
[Lemma('operate.v.01.operate'), Lemma('operate.v.01.run')]


In [ ]:
# Traversing the hieravhy of selected synset
# Get the synset for "cook"
cook_synset = wn.synset('run.v.4')

# Traverse the hypernym tree of "cook"
hypernyms = cook_synset.hypernyms()
while hypernyms:
    # Print the current synset and its definition
    print(f"{hypernyms[0].name()}")

    # Move up the tree to the next hypernym
    hypernyms = hypernyms[0].hypernyms()

direct.v.04
manage.v.02
control.v.01


With Verbs, the hierachy is more so a set of a related words that have similar meanings and it does not have a common top most level synset

In [ ]:
# Word to find different forms of
word = 'run'

# Get all possible forms of the word
forms = set()
for pos in ['n', 'v', 'a', 'r']:
    for form in wn._morphy(word, pos):
        forms.add(form)

# Print the different forms of the word
print(f"Different forms of '{word}':")
for form in forms:
    print(form)

Different forms of 'run':
run


Two words i think are similar

In [ ]:
# Word one
wn.synsets('throw')

[Synset('throw.n.01'),
 Synset('throw.n.02'),
 Synset('throw.n.03'),
 Synset('throw.n.04'),
 Synset('throw.n.05'),
 Synset('throw.v.01'),
 Synset('throw.v.02'),
 Synset('shed.v.01'),
 Synset('throw.v.04'),
 Synset('give.v.07'),
 Synset('throw.v.06'),
 Synset('project.v.10'),
 Synset('throw.v.08'),
 Synset('bewilder.v.02'),
 Synset('hurl.v.03'),
 Synset('hold.v.03'),
 Synset('throw.v.12'),
 Synset('throw.v.13'),
 Synset('throw.v.14'),
 Synset('confuse.v.02')]

In [ ]:
throw = wn.synset('throw.v.01')

In [ ]:
# Word two
wn.synsets('toss')


[Synset('flip.n.04'),
 Synset('pass.n.15'),
 Synset('toss.n.03'),
 Synset('flip.v.06'),
 Synset('flip.v.01'),
 Synset('chuck.v.01'),
 Synset('convulse.v.03'),
 Synset('discard.v.01'),
 Synset('toss.v.06')]

In [ ]:
chuck = wn.synset('chuck.v.01')

In [ ]:
# Wu-Palmer algo
wn.wup_similarity(throw, chuck)

0.8571428571428571

In [ ]:
# senses of `throw`
for ss in wn.synsets('throw'):
  print(ss, ss.definition())

Synset('throw.n.01') the act of throwing (propelling something with a rapid movement of the arm and wrist)
Synset('throw.n.02') a single chance or instance
Synset('throw.n.03') the maximum movement available to a pivoted or reciprocating piece by a cam
Synset('throw.n.04') bedclothes consisting of a lightweight cloth covering (an afghan or bedspread) that is casually thrown over something
Synset('throw.n.05') casting an object in order to determine an outcome randomly
Synset('throw.v.01') propel through the air
Synset('throw.v.02') move violently, energetically, or carelessly
Synset('shed.v.01') get rid of
Synset('throw.v.04') place or put with great energy
Synset('give.v.07') convey or communicate; of a smile, a look, a physical gesture
Synset('throw.v.06') cause to go on or to be engaged or set in operation
Synset('project.v.10') put or send forth
Synset('throw.v.08') to put into a state or activity hastily, suddenly, or carelessly
Synset('bewilder.v.02') cause to be confused emotion

In [ ]:
# Lesk algo
air = ["I", "can", "throw", "the", "ball", "farther", "than", "you", "can", "."]
print(lesk(air, 'throw'))


Synset('throw.v.14')


In [ ]:
# senses of `chuck`
for ss in wn.synsets('toss'):
  print(ss, ss.definition())

In [ ]:
air = ["She", "decided", "to", "toss", "the", "coin", "to", "make", "a", "decision", "."]
print(lesk(air, 'toss'))

Synset('flip.n.04')


Observation:
the synset `chuck.v.01` and `throw.v.01` scored a 0.86 using the Wu-Palmer algorithm. Which means, according to the algorithm, the synsets are very similar. 

The lesk algorithm got the wrong sense for `throw` it thought i was reffering to `throw.v.14` which is to throw a die on a surface rather than throwing an objet in the air.

The lesk algorith got the correct sense for toss however. it was able to figure out it meant flip a coin.

sentiWordNet is a lexical resource built on top of WordNet that assigns 3 sentiment scores for each synset: positivity, negativity and objectivity. You can use sentiWordNet to determine the ratio of positive to negative comments in a youtube comment section


In [ ]:
senti_list = list(swn.senti_synsets('bastard'))
for item in senti_list:
  print(item)

<position.n.03: PosScore=0.0 NegScore=0.0>
<view.n.02: PosScore=0.0 NegScore=0.0>
<view.n.03: PosScore=0.0 NegScore=0.0>
<view.n.04: PosScore=0.0 NegScore=0.0>
<opinion.n.01: PosScore=0.0 NegScore=0.625>
<opinion.n.02: PosScore=0.0 NegScore=0.5>
<view.n.07: PosScore=0.0 NegScore=0.0>
<scene.n.08: PosScore=0.0 NegScore=0.0>
<horizon.n.02: PosScore=0.125 NegScore=0.0>
<view.n.10: PosScore=0.0 NegScore=0.25>
<see.v.05: PosScore=0.125 NegScore=0.0>
<view.v.02: PosScore=0.25 NegScore=0.0>
<watch.v.03: PosScore=0.0 NegScore=0.0>


In [ ]:
import sys
from nltk.tokenize import word_tokenize

sentence = "Although the view from the mountain top was breathtakingly beautiful, the steep and rocky climb was treacherous and exhausting."
tokens = sentence.split()
neg, pos = 0, 0

for token in tokens:
  syn_list = list(swn.senti_synsets(token))
  for item in syn_list:
    neg += item.neg_score()
    pos += item.pos_score()

  print("Polarity of " + token + ". Positive: " + str(pos) + " Negative: " + str(neg))
  neg, pos = 0, 0
    
    
  

Polarity of Although. Positive: 0 Negative: 0
Polarity of the. Positive: 0 Negative: 0
Polarity of view. Positive: 0.5 Negative: 1.375
Polarity of from. Positive: 0 Negative: 0
Polarity of the. Positive: 0 Negative: 0
Polarity of mountain. Positive: 0.0 Negative: 0.0
Polarity of top. Positive: 0.125 Negative: 0.375
Polarity of was. Positive: 0.375 Negative: 0.25
Polarity of breathtakingly. Positive: 0 Negative: 0
Polarity of beautiful,. Positive: 0 Negative: 0
Polarity of the. Positive: 0 Negative: 0
Polarity of steep. Positive: 0.375 Negative: 0.75
Polarity of and. Positive: 0 Negative: 0
Polarity of rocky. Positive: 0.0 Negative: 1.75
Polarity of climb. Positive: 0.375 Negative: 0.0
Polarity of was. Positive: 0.375 Negative: 0.25
Polarity of treacherous. Positive: 0.125 Negative: 1.0
Polarity of and. Positive: 0 Negative: 0
Polarity of exhausting.. Positive: 0 Negative: 0


All of these scores makes no sense. Why is `view` given such a negative score? i check and it gave the senti synset of view, `opinion.n.01` a negative score of 0.625 and a 0.0 positive score. Why is `opinion` negative according to sentiWordNet. I also find it strange `breathtakingly` didnt get a positive score.  

In [ ]:
import sys
from nltk.tokenize import word_tokenize

sentence = "Although the view from the mountain top was breathtakingly beautiful, the steep and rocky climb was treacherous and exhausting."

neg = 0
pos = 0
tokens = sentence.split()
for token in tokens:
    syn_list = list(swn.senti_synsets(token))
    if syn_list:
        syn = syn_list[0]
        neg += syn.neg_score()
        pos += syn.pos_score()
    
print("neg\tpos counts")
print(neg, '\t', pos)

#***************************************************

neg	pos counts
0.625 	 0.0


sentence about collocation

In [ ]:
import nltk
from nltk.book import *
text4

In [ ]:
# get collocations
text4.collocations()

United States; fellow citizens; years ago; four years; Federal
Government; General Government; American people; Vice President; God
bless; Chief Justice; one another; fellow Americans; Old World;
Almighty God; Fellow citizens; Chief Magistrate; every citizen; Indian
tribes; public debt; foreign nations


In [ ]:
text = ' '.join(text4.tokens)
text[:66]

'Fellow - Citizens of the Senate and of the House of Representative'

In [ ]:
import math
vocab = len(set(text4))
hg= text.count('Fellow Citizens') / vocab
print("p(Fellow Citizens) = ", hg)
h = text.count('Fellow')/vocab
print("p(Fellow) = ", h)
g = text.count('Citizens')/vocab
print('p(Citizens) = ', g)
pmi = math.log2(hg / (h * g))
print('pmi = ', pmi)

p(Fellow Citizens) =  9.975062344139652e-05
p(Fellow) =  0.002394014962593516
p(Citizens) =  0.0006982543640897755
pmi =  5.898997193450885


In [ ]:
hg = text.count('of the')/vocab
print("p(of the) = ",hg )
o = text.count('of')/vocab
print("p(of) = ", o)
t = text.count('the ')/vocab # space so it doesn't capture 'their' etc.
print('p(the) = ', t)
pmi = math.log2(hg / (o * t))
print('pmi = ', pmi)

p(of the) =  0.20089775561097256
p(of) =  0.7487281795511221
p(the) =  0.9533167082294264
pmi =  -1.8290080938996587
